docs/Build.md

```diff
@@ -10,41 +10,67 @@ The design of the build system has the following main goals:
 
 * works identically on every platform
 * has minimal run-time dependencies:
-    - a working installation of `stack`
+    - `stack`
     - `git`
-* is completely functional right after simple `git clone`
-* one-stop-shop for building all executables required for using `hie` in IDEs.
+* is completely functional right after a simple `git clone` and after every `git pull`
+* one-stop-shop for building and naming all executables required for using `hie` in IDEs.
+* fails with meaningful error messages in known circumstances
+* is able to modify the environment such that `hie` can be run
+    - setup `hoogle` database
+    - setup `hlint` data-files
 
 See the project's `README` for detailed information about installing `hie`.
 
-### Tradeoffs
+### Targets
 
-#### `shake.yaml`
+The build script `install.hs` defines several targets using the `shake` build system. The targets are roughly:
 
-A `shake.yaml` is required for executing the `install.hs` file.
+* `hie-*`: builds and installs the `hie` binaries. Also renames the binaries to contain the correct version-number.
+* `build`: builds and installs `hie` binaries for all supported `ghc` versions.
+* `build-doc`: builds the hoogle-db required by `hie`
+* `cabal-*`: execute the same task as the original target, but with `cabal` instead of `stack`
 
-* It contains the required version of `shake`.
-* In contrast to the other `*.yaml` it does not contain the submodules, which is necessary for `stack` to work even before the submodules have been initialized.
+Each `stack-*.yaml` contains references to packages in the submodules. Calling `stack` with one of those causes the build to fail if the submodules have not been initialized already. The file `shake.yaml` solves this issue. Moreover, it specifies the correct version of `shake` and is used for installing all run-time dependencies such as `cabal` and `hoogle` if necessary.
```

Comment:
> The file `shake.yaml` solves this issue.

I think it's important to state that it solves the issue by running git binary itself (so no magic or Haskell-based libgit if such things exist).

Reply:
Good idea!